src/Compilers/CSharp/Portable/Binder/DecisionDagBuilder.cs

```diff
@@ -601,7 +884,69 @@ private static PartialCaseDecision RemoveEvaluation(PartialCaseDecision c, Bound
                 Index: c.Index,
                 Syntax: c.Syntax,
                 Decisions: c.Decisions.WhereAsArray(d => !(d is BoundDagEvaluation e2) || e2 != e),
-                Bindings: c.Bindings, WhereClause: c.WhereClause, CaseLabel: c.CaseLabel);
+                Bindings: c.Bindings, WhenClause: c.WhenClause, CaseLabel: c.CaseLabel);
+        }
+
+        private class DagState
+        {
+            public readonly ImmutableArray<PartialCaseDecision> Cases;
+
+            public DagState(ImmutableArray<PartialCaseDecision> cases)
+            {
+                this.Cases = cases;
+            }
+
+            public static implicit operator DagState(ImmutableArray<PartialCaseDecision> cases) => new DagState(cases);
+
+            // We only compute the dag states for the branches after we de-dup this DagState itself.
+            // If all that remains is the `when` clauses, SelectedDecision is left `null` and the
+            // FalseBranch field is populated with the succesor on failure of the when clause (if one exists).
+            public BoundDagDecision SelectedDecision;
+            public DagState TrueBranch, FalseBranch;
+
+            // After the entire graph of DatState objects is complete, we translate each into its Dag.
+            public BoundDecisionDag Dag;
+        }
+
+        /// <summary>
+        /// An equivalence relation between dag states used to dedup the states during dag construction.
+        /// After dag construction is complete we treat a DagState as using object equality as equivalent
+        /// states have been merged.
+        /// </summary>
+        private class DagStateEquivalence : IEqualityComparer<DagState>
+        {
+            public static readonly DagStateEquivalence Instance = new DagStateEquivalence();
+
+            public bool Equals(DagState x, DagState y)
+            {
+                ImmutableArray<PartialCaseDecision> xCases = x.Cases;
+                ImmutableArray<PartialCaseDecision> yCases = y.Cases;
+                if (xCases.Length != yCases.Length)
+                {
+                    return false;
+                }
+
+                for (int i = 0; i < xCases.Length; i++)
+                {
+                    if (xCases[i] != yCases[i])
+                    {
+                        return false;
+                    }
+                }
+
+                return true;
+            }
+
+            public int GetHashCode(DagState x)
+            {
+                int result = x.Cases.Length;
+                foreach (var c in x.Cases)
+                {
+                    result = Hash.Combine(c.GetHashCode(), result);
+                }
+
+                return result;
+            }
         }
 
         internal class PartialCaseDecision
```

Comment:
>internal [](start = 8, length = 8)

`private`?

Reply:
Yes.

---
In reply to: [172605248](https://github.com/dotnet/roslyn/pull/24059#discussion_r172605248) [](ancestors = 172605248)